File: server/src/services/jobParser.js
/**
 * Job description parser.
 * Extracts structured fields from raw job description text using NLP + regex.
 * Optionally uses OpenAI for richer extraction.
 * 
 * TODO: Convert to Python for better NLP capabilities (spaCy, NLTK)
 * TODO: Add validation to ensure input is a job description (not a resume)
 * TODO: Improve skill extraction accuracy with ML models
 */

import natural from 'natural';
import nlp from 'compromise';
import OpenAI from 'openai';

const { TfIdf, WordTokenizer } = natural;
const tokenizer = new WordTokenizer();

// ─── Skill taxonomy (same as nlpMatcher) ─────────────────────────────────────

const SKILL_GROUPS = {
  languages: ['javascript', 'typescript', 'python', 'java', 'c++', 'c#', 'ruby', 'go', 'golang', 'rust', 'swift', 'kotlin', 'scala', 'php', 'r', 'matlab', 'perl', 'bash', 'shell', 'powershell'],
  frontend:  ['react', 'angular', 'vue', 'svelte', 'next.js', 'nextjs', 'nuxt', 'html', 'css', 'sass', 'tailwind', 'bootstrap', 'webpack', 'vite', 'redux', 'graphql'],
  backend:   ['node', 'nodejs', 'express', 'fastapi', 'django', 'flask', 'spring', 'rails', 'laravel', 'nestjs', 'grpc', 'rest', 'microservices'],
  databases: ['sql', 'mysql', 'postgresql', 'postgres', 'mongodb', 'redis', 'elasticsearch', 'dynamodb', 'cassandra', 'oracle', 'sqlite', 'firebase'],
  cloud:     ['aws', 'azure', 'gcp', 'google cloud', 'docker', 'kubernetes', 'terraform', 'ansible', 'jenkins', 'ci/cd', 'devops', 'linux'],
  data:      ['machine learning', 'deep learning', 'nlp', 'tensorflow', 'pytorch', 'pandas', 'numpy', 'spark', 'hadoop', 'tableau', 'power bi'],
  practices: ['agile', 'scrum', 'kanban', 'tdd', 'git', 'jira', 'confluence', 'design patterns', 'microservices'],
  tools:     ['figma', 'sketch', 'salesforce', 'sap', 'excel', 'powerpoint'],
};
const ALL_SKILLS = Object.values(SKILL_GROUPS).flat();

const DEPARTMENTS = ['Engineering', 'Product', 'Design', 'Marketing', 'Sales', 'HR', 'Finance', 'Operations', 'Data', 'DevOps', 'QA', 'Security', 'Legal', 'Support'];
const JOB_TYPES   = ['Full-time', 'Part-time', 'Contract', 'Internship', 'Freelance', 'Remote'];

const EDUCATION_LEVELS = [
  { keywords: ['phd', 'ph.d', 'doctorate'],                                                label: "PhD / Doctorate",  score: 5 },
  { keywords: ['master', 'mba', 'msc', 'm.sc', 'ms ', 'm.s.'],                            label: "Master's",         score: 4 },
  { keywords: ['bachelor', 'bsc', 'b.sc', 'bs ', 'b.s.', 'b.e.', 'be ', 'undergraduate'], label: "Bachelor's",       score: 3 },
  { keywords: ['associate'],                                                                label: 'Associate',        score: 2 },
  { keywords: ['diploma'],                                                                  label: 'Diploma',          score: 1 },
  { keywords: ['certification', 'certified'],                                              label: 'Certification',    score: 1 },
];

const EXP_PATTERNS = [
  /(\d+)\+?\s*(?:to\s*\d+\s*)?years?\s+(?:of\s+)?(?:professional\s+)?(?:work\s+)?experience/gi,
  /experience\s*(?:of\s*)?(\d+)\+?\s*years?/gi,
  /(\d+)\+?\s*years?\s+(?:in|of|with)\s+\w/gi,
  /minimum\s+(?:of\s+)?(\d+)\s+years?/gi,
  /at\s+least\s+(\d+)\s+years?/gi,
];

// ─── NLP-based extraction ─────────────────────────────────────────────────────

function extractSkills(text) {
  const lower = text.toLowerCase();
  const found = new Set();
  
  // Extract known skills from taxonomy
  for (const skill of ALL_SKILLS) {
    const escaped = skill.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    if (new RegExp(`(?<![a-z])${escaped}(?![a-z])`, 'i').test(lower)) {
      found.add(skill);
    }
  }
  
  // Expanded stop words list to filter out common non-skill words
  const stopWords = new Set([
    // Articles, conjunctions, prepositions
    'The', 'And', 'For', 'With', 'This', 'That', 'From', 'Have', 'Will', 'Must', 'Should', 'Can', 'May',
    'Are', 'You', 'Our', 'Your', 'We', 'Be', 'Is', 'In', 'To', 'Of', 'At', 'As', 'An', 'Or', 'Not', 'But',
    'If', 'On', 'It', 'Do', 'No', 'So', 'Up', 'By', 'Us', 'My', 'He', 'She', 'They', 'Who', 'How', 'Why',
    'What', 'When', 'Where', 'Has', 'Had', 'Been', 'Being', 'Does', 'Did', 'Done', 'Would', 'Could',
    
    // Common job description words
    'Work', 'Team', 'Company', 'Project', 'Experience', 'Years', 'Role', 'Position', 'Job', 'Candidate',
    'Resume', 'Skills', 'Education', 'Degree', 'University', 'College', 'School', 'Responsibilities',
    'Requirements', 'Qualifications', 'Preferred', 'Required', 'Nice', 'Bonus', 'Plus', 'Ability',
    'Strong', 'Excellent', 'Good', 'Great', 'Solid', 'Proven', 'Demonstrated', 'Understanding', 'Knowledge',
    'Familiarity', 'Proficiency', 'Expertise', 'Competency', 'Capability', 'Skilled', 'Experienced',
    
    // Job levels and types
    'Senior', 'Junior', 'Lead', 'Manager', 'Engineer', 'Developer', 'Designer', 'Analyst', 'Architect',
    'Full', 'Part', 'Time', 'Remote', 'Hybrid', 'Onsite', 'Contract', 'Permanent', 'Temporary', 'Intern',
    'Entry', 'Mid', 'Level', 'Staff', 'Principal', 'Associate', 'Assistant', 'Coordinator', 'Specialist',
    
    // Action verbs
    'Develop', 'Build', 'Create', 'Design', 'Implement', 'Maintain', 'Support', 'Manage', 'Lead',
    'Collaborate', 'Communicate', 'Ensure', 'Provide', 'Deliver', 'Contribute', 'Participate', 'Assist',
    'Troubleshoot', 'Debug', 'Test', 'Deploy', 'Monitor', 'Optimize', 'Improve', 'Enhance', 'Integrate',
    
    // Location and logistics
    'Location', 'Office', 'Based', 'City', 'State', 'Country', 'Local', 'International', 'Global',
    
    // Misc common words
    'Including', 'Such', 'Other', 'Various', 'Multiple', 'Several', 'Many', 'Some', 'All', 'Any', 'Each',
    'Every', 'Both', 'Either', 'Neither', 'More', 'Most', 'Less', 'Least', 'Only', 'Just', 'Also', 'Even',
    'Well', 'Very', 'Too', 'Much', 'Few', 'Little', 'Large', 'Small', 'New', 'Old', 'High', 'Low',
    'Stay', 'Effectively', 'Efficiently', 'Successfully', 'Quickly', 'Rapidly', 'Fast', 'Slow',
  ]);
  
  // Extract capitalized words (likely to be tools/frameworks)
  const caps = text.match(/\b[A-Z][a-zA-Z0-9.+#]{2,}\b/g) || [];
  for (const w of caps) {
    if (!stopWords.has(w)) {
      const lw = w.toLowerCase();
      // Only add if it looks like a technical term
      const hasSpecialChars = /[0-9.+#]/.test(w);
      const isAcronym = w === w.toUpperCase() && w.length >= 2 && w.length <= 6;
      const isTechnicalPattern = /^[A-Z][a-z]+[A-Z]/.test(w); // CamelCase
      
      if (hasSpecialChars || isAcronym || isTechnicalPattern) {
        found.add(lw);
      }
    }
  }
  
  return [...found];
}

function extractExperience(text) {
  let maxYears = 0;
  for (const pattern of EXP_PATTERNS) {
    for (const m of [...text.matchAll(pattern)]) {
      const y = parseInt(m[1], 10);
      if (y > maxYears && y < 50) maxYears = y;
    }
  }
  return maxYears;
}

function extractEducation(text) {
  const lower = text.toLowerCase();
  for (const level of EDUCATION_LEVELS) {
    if (level.keywords.some((kw) => lower.includes(kw))) return level.label;
  }
  return null;
}

function extractTitle(text) {
  // Try first non-empty line that looks like a title (short, no period)
  const lines = text.split('\n').map((l) => l.trim()).filter(Boolean);
  for (const line of lines.slice(0, 5)) {
    if (line.length > 3 && line.length < 100 && !line.endsWith('.') && !line.startsWith('-')) {
      return line.replace(/^(job title|position|role)[:\s]*/i, '').trim();
    }
  }
  return '';
}

function extractDepartment(text) {
  const lower = text.toLowerCase();
  for (const dept of DEPARTMENTS) {
    if (lower.includes(dept.toLowerCase())) return dept;
  }
  return 'Engineering';
}

function extractJobType(text) {
  const lower = text.toLowerCase();
  for (const type of JOB_TYPES) {
    if (lower.includes(type.toLowerCase())) return type;
  }
  return 'Full-time';
}

function extractLocation(text) {
  const doc = nlp(text);
  const places = doc.places().out('array');
  // Filter out generic words
  const filtered = places.filter((p) => p.length > 2 && p.length < 50);
  return filtered[0] || '';
}

function extractNiceToHave(text) {
  const lower = text.toLowerCase();
  const niceSection = lower.match(/(?:nice[- ]to[- ]have|preferred|bonus|plus|advantageous)[:\s\n]+([\s\S]{0,500}?)(?:\n\n|\n[A-Z]|$)/i);
  if (!niceSection) return [];
  const sectionText = niceSection[1];
  return extractSkills(sectionText);
}

function extractResponsibilities(text) {
  // Extract bullet points that look like responsibilities
  const bullets = text.match(/^[\s]*[-•*]\s+(.+)$/gm) || [];
  return bullets
    .map((b) => b.replace(/^[\s]*[-•*]\s+/, '').trim())
    .filter((b) => b.length > 10 && b.length < 200)
    .slice(0, 10);
}

function extractSalary(text) {
  const salaryMatch = text.match(/\$[\d,]+(?:\s*[-–]\s*\$[\d,]+)?(?:\s*(?:per\s+)?(?:year|yr|annual|month|hour|hr))?/i)
    || text.match(/[\d,]+k?\s*[-–]\s*[\d,]+k?\s*(?:USD|GBP|EUR)?/i);
  return salaryMatch ? salaryMatch[0].trim() : '';
}

/**
 * Parse a job description using NLP (no API key needed).
 */
export function parseJobNLP(text) {
  const skills        = extractSkills(text);
  const niceToHave    = extractNiceToHave(text);
  const requiredSkills = skills.filter((s) => !niceToHave.includes(s));

  return {
    title:            extractTitle(text),
    department:       extractDepartment(text),
    location:         extractLocation(text),
    type:             extractJobType(text),
    minExperience:    extractExperience(text),
    educationLevel:   extractEducation(text),
    requiredSkills,
    niceToHaveSkills: niceToHave,
    responsibilities: extractResponsibilities(text),
    salary:           extractSalary(text),
    rawText:          text,
    parsedBy:         'nlp',
  };
}

/**
 * Parse a job description using OpenAI for richer structured extraction.
 */
export async function parseJobAI(text) {
  if (!process.env.OPENAI_API_KEY) return parseJobNLP(text);

  const client = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });
  const model  = 'gpt-4o'; // Use GPT-4o (5.2) for best parsing quality

  const prompt = `You are an expert HR analyst. Extract structured information from this job description.

JOB DESCRIPTION:
${text.slice(0, 4000)}

IMPORTANT INSTRUCTIONS FOR SKILLS EXTRACTION:
- "requiredSkills" and "niceToHaveSkills" should ONLY include actual technical skills, tools, frameworks, programming languages, and technologies
- DO NOT include: job duties (e.g., "troubleshoot", "develop"), soft skills (e.g., "communication"), qualifications (e.g., "preferred", "required"), adjectives (e.g., "good", "strong", "understanding"), location terms, or common words
- Examples of VALID skills: "JavaScript", "React", "Python", "AWS", "Docker", "SQL", "Kubernetes", "TypeScript", "Node.js", "MongoDB"
- Examples of INVALID skills: "troubleshoot", "understanding", "preferred", "required", "full", "location", "why", "effectively", "stay", "proficiency", "familiarity", "good"
- Only include concrete, specific technical competencies that can be learned and demonstrated

Respond with ONLY valid JSON:
{
  "title": "<job title>",
  "department": "<one of: Engineering, Product, Design, Marketing, Sales, HR, Finance, Operations, Data, DevOps, QA, Security, Legal, Support, Other>",
  "location": "<city/remote/hybrid or empty string>",
  "type": "<one of: Full-time, Part-time, Contract, Internship, Freelance, Remote>",
  "minExperience": <number of years, 0 if not specified>,
  "educationLevel": "<Bachelor's | Master's | PhD / Doctorate | Associate | Diploma | Certification | null>",
  "requiredSkills": ["skill1", "skill2"],
  "niceToHaveSkills": ["skill1", "skill2"],
  "responsibilities": ["responsibility1", "responsibility2"],
  "salary": "<salary range or empty string>",
  "summary": "<2-3 sentence summary of the role>"
}`;

  try {
    const response = await client.chat.completions.create({
      model,
      messages: [{ role: 'user', content: prompt }],
      temperature: 0,  // Set to 0 for maximum consistency
      max_tokens: 1000,
      seed: 12345,     // Fixed seed for reproducible results
      response_format: { type: 'json_object' },
    });

    const parsed = JSON.parse(response.choices[0].message.content);
    return { ...parsed, rawText: text, parsedBy: 'ai' };
  } catch (err) {
    console.error('AI job parse failed, falling back to NLP:', err.message);
    return parseJobNLP(text);
  }
}

File: server_py/services/job_parser.py
"""
services/job_parser.py — Job description structured extraction.

Two parsing modes:
- parse_job_nlp()  — fully offline, regex + keyword matching against a skill
  taxonomy; extracts title, department, location, type, experience, education,
  required/nice-to-have skills, responsibilities, and salary.
- parse_job_ai()   — sends the description to GPT-4o with a strict JSON prompt
  for higher-accuracy extraction; falls back to NLP on failure.

Both return the same dict shape so callers are agnostic to which was used.
"""

import os
import re
import json
from openai import AsyncOpenAI
import anthropic

SKILL_GROUPS = {
    "languages": ["javascript","typescript","python","java","c++","c#","ruby","go","golang","rust","swift","kotlin","scala","php","r","matlab","perl","bash","shell","powershell"],
    "frontend":  ["react","angular","vue","svelte","next.js","nextjs","nuxt","html","css","sass","tailwind","bootstrap","webpack","vite","redux","graphql"],
    "backend":   ["node","nodejs","express","fastapi","django","flask","spring","rails","laravel","nestjs","grpc","rest","microservices"],
    "databases": ["sql","mysql","postgresql","postgres","mongodb","redis","elasticsearch","dynamodb","cassandra","oracle","sqlite","firebase"],
    "cloud":     ["aws","azure","gcp","google cloud","docker","kubernetes","terraform","ansible","jenkins","ci/cd","devops","linux"],
    "data":      ["machine learning","deep learning","nlp","tensorflow","pytorch","pandas","numpy","spark","hadoop","tableau","power bi"],
    "practices": ["agile","scrum","kanban","tdd","git","jira","confluence","design patterns","microservices"],
}
ALL_SKILLS = [s for g in SKILL_GROUPS.values() for s in g]

DEPARTMENTS = ["Engineering","Product","Design","Marketing","Sales","HR","Finance","Operations","Data","DevOps","QA","Security","Legal","Support"]
JOB_TYPES   = ["Full-time","Part-time","Contract","Internship","Freelance","Remote"]

EDUCATION_LEVELS = [
    (["phd","ph.d","doctorate"],            "PhD / Doctorate"),
    (["master","mba","msc","m.sc","ms "],   "Master's"),
    (["bachelor","bsc","b.sc","bs ","b.e."],"Bachelor's"),
    (["associate"],                         "Associate"),
    (["diploma"],                           "Diploma"),
    (["certification","certified"],         "Certification"),
]

EXP_PATTERNS = [
    re.compile(r"(\d+)\+?\s*(?:to\s*\d+\s*)?years?\s+(?:of\s+)?(?:professional\s+)?(?:work\s+)?experience", re.I),
    re.compile(r"experience\s*(?:of\s*)?(\d+)\+?\s*years?", re.I),
    re.compile(r"(\d+)\+?\s*years?\s+(?:in|of|with)\s+\w", re.I),
    re.compile(r"minimum\s+(?:of\s+)?(\d+)\s+years?", re.I),
    re.compile(r"at\s+least\s+(\d+)\s+years?", re.I),
]


def _extract_skills(text: str) -> list[str]:
    lower = text.lower()
    found: set[str] = set()
    for skill in ALL_SKILLS:
        if re.search(rf"(?<![a-z]){re.escape(skill)}(?![a-z])", lower, re.I):
            found.add(skill)
    stop = {"The","And","For","With","Work","Team","Company","Senior","Junior","Lead","Manager","Engineer","Developer"}
    for w in re.findall(r"\b[A-Z][a-zA-Z0-9.+#]{2,}\b", text):
        if w not in stop:
            lw = w.lower()
            if re.search(r"[0-9.+#]", w) or (w == w.upper() and 2 <= len(w) <= 6) or re.match(r"^[A-Z][a-z]+[A-Z]", w):
                found.add(lw)
    return list(found)


def _extract_experience(text: str) -> int:
    best = 0
    for pat in EXP_PATTERNS:
        for m in pat.finditer(text):
            y = int(m.group(1))
            if y > best and y < 50:
                best = y
    return best


def _extract_education(text: str):
    lower = text.lower()
    for keywords, label in EDUCATION_LEVELS:
        if any(kw in lower for kw in keywords):
            return label
    return None


def _extract_title(text: str) -> str:
    for line in text.split("\n")[:5]:
        line = line.strip().lstrip("#").strip()  # strip markdown heading markers
        if 3 < len(line) < 100 and not line.endswith(".") and not line.startswith("-"):
            return re.sub(r"^(job title|position|role)[:\s]*", "", line, flags=re.I).strip()
    return ""


def _extract_department(text: str) -> str:
    lower = text.lower()
    for d in DEPARTMENTS:
        if d.lower() in lower:
            return d
    return "Engineering"


def _extract_job_type(text: str) -> str:
    lower = text.lower()
    for t in JOB_TYPES:
        if t.lower() in lower:
            return t
    return "Full-time"


def _extract_location(text: str) -> str:
    m = re.search(r"\b(?:location|based in|office)[:\s]+([A-Za-z ,]+)", text, re.I)
    return m.group(1).strip()[:50] if m else ""


def _extract_nice_to_have(text: str) -> list[str]:
    m = re.search(r"(?:nice[- ]to[- ]have|preferred|bonus|plus|advantageous)[:\s\n]+([\s\S]{0,500}?)(?:\n\n|\n[A-Z]|$)", text, re.I)
    return _extract_skills(m.group(1)) if m else []


def _extract_responsibilities(text: str) -> list[str]:
    bullets = re.findall(r"^[\s]*[-•*]\s+(.+)$", text, re.MULTILINE)
    return [b.strip() for b in bullets if 10 < len(b.strip()) < 200][:10]


def _extract_salary(text: str) -> str:
    m = re.search(r"\$[\d,]+(?:\s*[-–]\s*\$[\d,]+)?(?:\s*(?:per\s+)?(?:year|yr|annual|month|hour|hr))?", text, re.I)
    return m.group(0).strip() if m else ""


def parse_job_nlp(text: str) -> dict:
    skills = _extract_skills(text)
    nice = _extract_nice_to_have(text)
    required = [s for s in skills if s not in nice]
    return {
        "title":            _extract_title(text),
        "department":       _extract_department(text),
        "location":         _extract_location(text),
        "type":             _extract_job_type(text),
        "minExperience":    _extract_experience(text),
        "educationLevel":   _extract_education(text),
        "requiredSkills":   required,
        "niceToHaveSkills": nice,
        "responsibilities": _extract_responsibilities(text),
        "salary":           _extract_salary(text),
        "rawText":          text,
        "parsedBy":         "nlp",
    }


def _get_provider() -> str:
    if os.getenv("OPENAI_API_KEY"):
        return "openai"
    if os.getenv("ANTHROPIC_API_KEY"):
        return "anthropic"
    return ""


_JOB_PARSE_PROMPT = """You are an expert HR analyst. Extract structured information from this job description.

JOB DESCRIPTION:
{text}

IMPORTANT: requiredSkills and niceToHaveSkills must ONLY be actual technical skills/tools/frameworks/languages.
DO NOT include: job duties, soft skills, adjectives, or common words.

Respond with ONLY valid JSON:
{{
  "title": "<job title>",
  "department": "<Engineering|Product|Design|Marketing|Sales|HR|Finance|Operations|Data|DevOps|QA|Security|Legal|Support|Other>",
  "location": "<city/remote/hybrid or empty string>",
  "type": "<Full-time|Part-time|Contract|Internship|Freelance|Remote>",
  "minExperience": <number>,
  "educationLevel": "<Bachelor's|Master's|PhD / Doctorate|Associate|Diploma|Certification|null>",
  "requiredSkills": ["skill1"],
  "niceToHaveSkills": ["skill1"],
  "responsibilities": ["responsibility1"],
  "salary": "<salary range or empty string>",
  "summary": "<2-3 sentence summary>"
}}"""


async def parse_job_ai(text: str) -> dict:
    provider = _get_provider()
    if not provider:
        return parse_job_nlp(text)

    prompt = _JOB_PARSE_PROMPT.format(text=text[:4000])
    try:
        if provider == "openai":
            client = AsyncOpenAI(api_key=os.getenv("OPENAI_API_KEY"))
            resp = await client.chat.completions.create(
                model=os.getenv("OPENAI_AI_MODEL", "gpt-4o-mini"),
                messages=[{"role": "user", "content": prompt}],
                temperature=float(os.getenv("AI_TEMPERATURE", "0.2")),
                max_tokens=int(os.getenv("AI_MAX_TOKENS", "1500")),
                seed=12345,
                response_format={"type": "json_object"},
            )
            raw = resp.choices[0].message.content
        else:
            client = anthropic.AsyncAnthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))
            resp = await client.messages.create(
                model=os.getenv("AI_MODEL", "claude-haiku-4-5-20251001"),
                max_tokens=int(os.getenv("AI_MAX_TOKENS", "1500")),
                temperature=float(os.getenv("AI_TEMPERATURE", "0.2")),
                messages=[{"role": "user", "content": prompt}],
            )
            raw = resp.content[0].text.strip()
            if "```" in raw:
                raw = raw.split("```")[1]
                if raw.startswith("json"):
                    raw = raw[4:]
        parsed = json.loads(raw)
        return {**parsed, "rawText": text, "parsedBy": "ai"}
    except Exception as e:
        print(f"AI job parse failed, falling back to NLP: {e}")
        return parse_job_nlp(text)
